### PR TITLE
Add gateway route table routes logic

### PR DESCRIPTION
**Pull request by mortenlerudjordet**
## Description

Adds support for gateway route table routes.
Default route added is for gw subnet range to azure firewall internal ip.

Also support for adding more routes through route input variables:

- If only address_prefix is defined, the route will be set to point at the internal ip of the azure firewall
- If address_prefix, next_hop_type and next_hop_in_ip_address, the value of these will be used instead
- Can control if default route to fw is added or not setting route_table_gw_fw_route_enabled, default is true

Fixes Azure/Azure-Landing-Zones#339

## Type of Change

- [ ] Non-module change (e.g. CI/CD, documentation, etc.)
- [x] Azure Verified Module updates:
  - [x] Bugfix containing backwards compatible bug fixes
    - [x] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [x] Feature update backwards compatible feature updates.
  - [ ] Breaking changes.
  - [x] Update to documentation

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] I did run all  [pre-commit](https://azure.github.io/Azure-Verified-Modules/contributing/terraform/terraform-contribution-flow/#5-run-pre-commit-checks) checks

#### Both type of routes
image

#### Custom route and default disabled
image

#### Both type of routes
image

#### Custom route and default disabled
image

#### Both type of routes
image

#### Custom route and default disabled
image

#### Both type of routes
image

#### Custom route and default disabled
image

### CODE REVIEW

**Comment by mortenlerudjordet** [comment]
Ran into an issue with first plan before gw subnet exists that I need to figure out a solution for before this PR is ready.

image

**Comment by mortenlerudjordet** [comment]
need more testing, will open another PR once I get it all sorted

**Comment by mortenlerudjordet** [comment]
@jaredfholgate, solved the issues with terraform I was having on plan. Made sure the code is as in line with the other logic and made as few changes as I was able to on the existing code state to get custom routes to work for gw.

File: variables.tf [hunk 1]
```diff
@@ -348,6 +348,13 @@ variable "hub_virtual_networks" {
       route_table_creation_enabled              = optional(bool, false)
       route_table_name                          = optional(string)
       route_table_bgp_route_propagation_enabled = optional(bool, false)
+      route_table_gw_fw_route_enabled           = optional(bool, true)
```
Comment (jaredfholgate) [inline variables.tf]: Can we avoid the acronyms here and replace `gw` and `fw` with the full words.

File: main.tf [hunk 2]
```diff
@@ -60,7 +59,12 @@ module "gateway_route_table" {
   resource_group_name           = each.value.resource_group_name
   bgp_route_propagation_enabled = each.value.bgp_route_propagation_enabled
   enable_telemetry              = var.enable_telemetry
+  routes                        = each.value.routes
+  subnet_resource_ids           = each.value.subnet_resource_ids
   tags                          = var.tags
+
+  # to stop api contention on gw subnet make sure route table is created after gw is finished
+  depends_on = [module.virtual_network_gateway]
```
Comment (jaredfholgate) [inline main.tf]: We avoid explicit depends on for modules as it is an anti-pattern that can have unexpected consequences.

**Review by jaredfholgate — CHANGES_REQUESTED** [review]
I think these changes will impact day 2 operations and module idempotency. Can you take another look. I can help if you are seeing another issue related to this.

File: main.tf [hunk 3]
```diff
@@ -60,7 +59,12 @@ module "gateway_route_table" {
   resource_group_name           = each.value.resource_group_name
   bgp_route_propagation_enabled = each.value.bgp_route_propagation_enabled
   enable_telemetry              = var.enable_telemetry
+  routes                        = each.value.routes
+  subnet_resource_ids           = each.value.subnet_resource_ids
```
[the remark this answers is not in the document]
Comment (mortenlerudjordet) [reply main.tf]: The local should handle validating the values, and set both to what the default values set by the avm sub module would be if they where omitted.

File: locals.subnets.tf [hunk 4]
```diff
@@ -17,7 +17,7 @@ locals {
       address_prefixes = [coalesce(value.virtual_network_gateways.subnet_address_prefix, local.virtual_network_subnet_default_ip_prefixes[key]["gateway"])]
       name             = "GatewaySubnet"
       route_table = {
-        id                           = local.gateway_route_table_enabled[key] ? module.gateway_route_table[key].resource_id : null
+        id                           = null
```
[the remark this answers is not in the document]
Comment (mortenlerudjordet) [reply locals.subnets.tf]: If we remove setting subnet_resource_ids from the route module and have this done in the subnet code, we still would need to make sure the route table be created before the subnet creation logic runs in the submodule. Else it will complain during apply that it can not associate the route table as it has not been created yet (happened on several of the tests I have run). 
My take is the route creation code can not stay where it is at for this to work.

I will do some more testing as I think i clearer understand the two issues at work here that throws up problems for terraform.

[the remark this answers is not in the document]
Comment (mortenlerudjordet) [reply locals.subnets.tf]: So after having a think and connecting all the dots in my head, I saw the flaws in my initial approach to this.
Did a refactor of the logic with this new understanding of the underlying issues, and ended up using the same approach as for the other existing route table creation logic.

So now the logic is split into 3 parts:
1. Create empty gw route table (as before the change)
2. Let the subnet logic in the submodule assign the empty route table to the gw subnet during creation (as before)
3. Add in the calculated routes in a separate resource block that uses the output from module.hub_and_spoke_vnet to get the fw private ip address (as done for the other route table creation logic).

After running test where I selective turn on and off the different switches for gw creation, gw route table creation, adding and removing all custom routes and enable/disable creation of default routes it seems to do as expected. Though had to add some extra test logic to add empty map/list elements based on if different components are enabled or not. This helps with handling some edge cases where the plan would throw errors for using empty values/map/objects.

Also 2 apply now does not remove the gw route table from the gw subnet.

Tok a couple of rounds to get there, but the end result should now be good.

**Comment by mortenlerudjordet** [comment]
@jaredfholgate : did a refactor of the logic and adopted the same method as used for the other route creation logic where gw route table creation is split into 3 separate module/resource blocks. This should solve the issues you brought up in the review.

Did need to sit down and connect all the dots in my head before I understood all the nuances at play here.
Ran many extra test iteration on plan and apply (and 2. apply runs) to see if I could shake out any more edge cases.
